include/nbl/builtin/hlsl/mpl.hlsl

```diff
@@ -69,47 +54,72 @@ uint64_t asuint64_t<float64_t>(float64_t val)
 
-template<uint64_t N>
+namespace impl
+{
+
+template<uint64_t N, uint16_t bits>
 struct countl_zero
 {
-    NBL_CONSTEXPR_STATIC_INLINE uint16_t value =
-#ifdef __HLSL_VERSION
-      impl::countl_zero<N, 6>::value;
+    NBL_CONSTEXPR_STATIC_INLINE uint64_t SHIFT = bits >> 1;
+    NBL_CONSTEXPR_STATIC_INLINE uint64_t LO_MASK = (1ull << SHIFT) - 1;
+    NBL_CONSTEXPR_STATIC_INLINE bool CHOOSE_HIGH = N & (LO_MASK << SHIFT);
+    NBL_CONSTEXPR_STATIC_INLINE uint64_t NEXT = (CHOOSE_HIGH ? (N >> SHIFT) : N) & LO_MASK;
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t value = countl_zero<NEXT, SHIFT>::value + (CHOOSE_HIGH ? 0ull : SHIFT);
+};
+
+template<uint64_t N>
+struct countl_zero<N, 1> : integral_constant<uint16_t, uint16_t(1u - (N & 1))>
+{};
+
+}
+
+template<class T, T N>
+struct countl_zero : impl::countl_zero<uint64_t(N), (sizeof(T) * 8)>
+{
+#ifdef __cplusplus
+    static_assert
 #else
-      std::countl_zero(N);
+    _Static_assert
 #endif
+    (is_integral<T>::value, "countl_zero type parameter must be an integral type");
 };
 
 template<uint64_t X>
 struct log2
 {
-    NBL_CONSTEXPR_STATIC_INLINE uint16_t value = X ? (1ull<<6)-countl_zero<X>::value-1 : -1ull;
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t value = X ? (1ull<<6)-countl_zero<uint64_t, X>::value-1 : -1ull;
 };
 
 
 #ifdef __HLSL_VERSION
-template<class T, class U = T>
-T bit_cast(U val);
+// template<class T, class U = T>
+// T bit_cast(U val);
+
+// #define NBL_DECLARE_BIT_CAST(FROM, TO) template<> TO bit_cast<TO, FROM>(FROM val) { return mpl::impl::as##TO (val); }
 
-#define NBL_DECLARE_BIT_CAST(FROM, TO) template<> TO bit_cast<TO, FROM>(FROM val) { return mpl::impl::as##TO (val); }
+// #define NBL_DECLARE_BIT_CAST_TYPES(BASE, BITS) \
+//     NBL_DECLARE_BIT_CAST(BASE ## BITS, float ## BITS) \
+//     NBL_DECLARE_BIT_CAST(BASE ## BITS, uint ## BITS) \
+//     NBL_DECLARE_BIT_CAST(BASE ## BITS, int ## BITS)
 
-#define NBL_DECLARE_BIT_CAST_TYPES(BASE, BITS) \
-    NBL_DECLARE_BIT_CAST(BASE ## BITS, float ## BITS) \
-    NBL_DECLARE_BIT_CAST(BASE ## BITS, uint ## BITS) \
-    NBL_DECLARE_BIT_CAST(BASE ## BITS, int ## BITS)
+// #define NBL_DECLARE_BIT_CAST_BITS(BASE) \
+//     NBL_DECLARE_BIT_CAST_TYPES(BASE, 16_t) \
+//     NBL_DECLARE_BIT_CAST_TYPES(BASE, 32_t) \
+//     NBL_DECLARE_BIT_CAST_TYPES(BASE, 64_t)
 
-#define NBL_DECLARE_BIT_CAST_BITS(BASE) \
-    NBL_DECLARE_BIT_CAST_TYPES(BASE, 16_t) \
-    NBL_DECLARE_BIT_CAST_TYPES(BASE, 32_t) \
-    NBL_DECLARE_BIT_CAST_TYPES(BASE, 64_t)
+// NBL_DECLARE_BIT_CAST_BITS(float)
+// NBL_DECLARE_BIT_CAST_BITS(uint)
+// NBL_DECLARE_BIT_CAST_BITS(int)
 
-NBL_DECLARE_BIT_CAST_BITS(float)
-NBL_DECLARE_BIT_CAST_BITS(uint)
-NBL_DECLARE_BIT_CAST_BITS(int)
+// #undef NBL_DECLARE_BIT_CAST
+// #undef NBL_DECLARE_BIT_CAST_TYPES
+// #undef NBL_DECLARE_BIT_CAST_BITS
 
-#undef NBL_DECLARE_BIT_CAST
-#undef NBL_DECLARE_BIT_CAST_TYPES
-#undef NBL_DECLARE_BIT_CAST_BITS
+template<class T, class U>
+T bit_cast(U val)
+{
+    return spirv::bitcast<T, U>(val);
```

Comment:
you probably want to static-assert `sizeof(T)<=sizeof(U)`
(but comment it out for now cause DXC )